Find Best Path Using Constraints

MarvinTorres opened the pull request:
**IMPORTANT: See [this post](https://github.com/kytos/pathfinder/pull/58#issuecomment-664761287) before merging!**

This updates Pathfinder to support searching for constrained shortest flexible paths via REST API calls to the main module.

Note that the changes are so extensive that tests created for the old graph module might fail.

**Overview**

_Note_: Tests that invoked KytosGraph's _set_default_metadata() will fail, because it was removed as part of our changes. 

KytosGraph has been extended to support searches for constrained flexible paths using graphs with edge attribute metadata. 

Edges can now contain the following attributes:

- Priority
- Utilization
- Reliability
- Ownership
- Delay
- Bandwidth

It also uses six Filters (one for each attribute) that produce subgraphs, which have edges that meet user-specified attribute requirements, such as "Bandwidth must be 20 or more."

In addition, users can specify flexible ("would like to have") requirements on top of the base ("must have") requirements. 

**Implementation**

[REST API Documentation](https://documenter.getpostman.com/view/9650761/T1Dngd4D)

Method Signature:
`constrained_flexible_paths(source, destination, depth=None, **metrics)`

Filters

Six filters from a new Filter class are now used by KytosGraph. Each attribute maps to one Filter. Each Filter contains a required type for the value and one of three filtering functions.

When KytosGraph creates a subgraph, it picks a user-specified attribute name ("bandwidth", "delay", etc.) from the user-specified requirement list, finds the attribute's associated Filter (one of three, explained below). Then it calls that Filter's run method, which accepts a graph and user-specified attribute value and returns a subgraph. If it has more attributes to process, then it repeats everything but with new attribute names and values.

Filtering Functions

_FilterLEQ_ Lower is better. Edges must be at most a user-specified value.
- Utilization
- Delay

_FilterGEQ_ Higher is better. Edges must be at least a user-specified value.
- Bandwidth
- Priority
- Reliability

_FilterEEQ_ Must match exactly. Edges must match a user-specified value.
- Ownership

Usage Examples:
```
1. constrained_flexible_paths("S1", "S1")
2. constrained_flexible_paths("S1", "S3", base={"bandwidth": 50})
3. constrained_flexible_paths("S1", "S2", 4, base={"ownership": "B"}, flexible={"delay": 50, "bandwidth": 100, "reliability": 3})
```
1 returns source, since source and destination are the same.
2 returns the shortest path from S1 to S3 with edges that have at least 50 bandwidth.
3 returns the shortest path from S1 to S2 with edges that are owned by B, with the possibility of those edges having at most 50 delay, at least 100 bandwidth, or at least 3 reliability.

Base and flexible metrics are represented as two dictionaries. To pass those dictionaries to the constrained shortest path function, use the keyword arguments _base_ for the base requirements dictionary and _flexible_ for the flexible requirements dictionary.

Main has been updated to reflect these changes.

Three test suites have been provided.

constrained_flexible_paths(...) returns a list of dictionaries:

`[{"paths" : <set of edges from shortest path>, "metrics" : <set of fulfilled requirements> }, ...]`

Each dictionary represents a shortest path, and the user-specified attributes used to find it. The base requirements are guaranteed to be fulfilled, while the flexible requirements may or may not be fulfilled.

**Algorithm for constrained_flexible paths**

```
1. If source is in the graph and source == destination, return source
2. Create subgraph A using the edges in the graph that fulfill the _base_ requirements.
3. Create "power set" P of combinations of _flexible_ requirements. Note that it's not a true power set since it excludes the empty set.
4. For each combination p(i) in P, where 1 < i < |P|
5.    Create subgraph A* using the edges from A that fulfill p(i).
6.    Find the shortest path from source to destination in A*.
7.    If at least one shortest path found, break from the for loop and return results.
```
- This assumes full depth (6). Depth affects the size of P - If it is less than 6 (the current max of the attribute count) then P will have fewer combinations.
- In addition, this algorithm's complexity grows exponentially based on the number of attributes, since the size of P doubles for every added attribute.

**Class Diagram for Main, KytosGraph, and Filter**

![class](https://user-images.githubusercontent.com/26232289/85330898-30d29f00-b4a3-11ea-82c3-711fe8c138f0.png)
**Sequence Diagram for constrained_flexible paths**

![sequence](https://user-images.githubusercontent.com/26232289/85330928-39c37080-b4a3-11ea-8080-ef932bb5c588.png)

Ktmi committed:
Added start of integration tests

Ktmi committed:
Switching to unit tests.

Ktmi committed:
test_setup prints the list of nodes and edges.

Ktmi committed:
Adds new tests

Ktmi committed:
Modified test methods for more info collection.

Ktmi committed:
Added in basics for constrained shortest paths

Ktmi committed:
Nearly finished filters, need to define tests

Ktmi committed:
Fixes error with updating default metadata

Ktmi committed:
Removes _set_default_metadata method

Ktmi committed:
Updated tests for constrained shortest path

Ktmi committed:
identified and fixed a problem with pathing

Ktmi committed:
Added a crude working implementation of flexible

Ktmi committed:
Improved code efficiency for brute force method

Ktmi committed:
Added in  new graph to tests

Ktmi committed:
Moved new graph for into seperate test driver

Ktmi committed:
Several changes to graph and tests.
 - Graph now allows for selection of algorithm.
 - Default graph algorithm is now networkx.all_shortest_paths
 - Updated tests to change algorithm to networkx.shortest_simple_paths
 - Fixed capitilization error in test_graph2

Commit:
5 unit tests skeleton

Ktmi committed:
More tests & a single bug fix
 - Tests for the following constraints
   + reliability
   + delay
   + bandwidth
 - Fixed missing dictionary entry for reliability

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
added test_graph2.py

Commit:
added unit test skeleton

Commit:
working on test 4

Ktmi committed:
Added new tests
 - Multiple constraints at once
 - Multiple constraints with flexibility

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
added new graph test for self

Ktmi committed:
Multiple changes
 - Changed constrained_flexible_paths
   + Flexible is now an integer instead of a bool
   + Can declare levels of flexibility, 0 being none
 - Added new rest endpoint for constrained flexible paths
 - Updated tests to reflect changes to constrained flexible

Commit:
merged with David's changes

Commit:
completed test graph

Ktmi committed:
Fixed flexibility in test_graph2

Ktmi committed:
Merge changes from Ktmi fork into main repository
Changes include:
 - Added new method for pathfinding, based on constrained shortest path
 - Added REST endpoint for new pathfinding method
 - Added several tests for graph.py, for both old and new methods

Ktmi committed:
Remove unnecessary test_pathfinder

Commit:
update test_graph3.py

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
update test_graph3.py

Commit:
added second test case

Ktmi committed:
REST Enpoint updates
 - Fixes errors in REST endpoint
 - Adds support to constrained shortest path to webui

Commit:
finished unit tests

Ktmi committed:
Finished webui
 + Adds in support for constraints
 + Adds in support for flexibility

Ktmi committed:
Removed code from web ui used to help debug

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
test_graph3 now subclass of test_graph

Commit:
test_graph2 changed back to David's code

Commit:
test_graph3 class name now consistent

Ktmi committed:
Merge pull request #1 from MarvinTorres/master

Marvin Torres's changes

Ktmi committed:
Revert "Marvin Torres's changes"

Ktmi committed:
Merge pull request #2 from Ktmi/revert-1-master

Revert "Marvin Torres's changes"

Ktmi committed:
Reduced verbosity of tests

Commit:
test_graph3 removed redundant methods

Commit:
Merge branch 'master' of https://github.com/Ktmi/pathfinder

Commit:
fixes for test_graph3 and graph

Commit:
test_graph3 less verbose

Ktmi committed:
Merge pull request #4 from MarvinTorres/master

Addition of test_graph3.py and minor change to graph.py

Ktmi committed:
Merge remote-tracking branch 'upstream/master'

Ktmi committed:
New filter system, throws error for wrong type

Ktmi committed:
Added in new way to do flexibility

Ktmi committed:
Modifies rest endpoints.

Ktmi committed:
added brakes to constrained flexible

Ktmi committed:
Modifies pathfinder to reduce duplicate code

Ktmi committed:
Added test for type checking

Ktmi committed:
Final changes

Ktmi committed:
Slight speed optimization

Commit:
fixed test_graph3.py

Ktmi committed:
Merge pull request #5 from MarvinTorres/test3fixes

Fixed test_graph3.py

Commit:
fixed graph setup

Ktmi committed:
Merge pull request #6 from MarvinTorres/test3fixes

fixed graph setup

Commit:
fixed test_path9

Ktmi committed:
Merge pull request #7 from MarvinTorres/test3fixes

fixed test_path9

MarvinTorres committed:
removed trailing whitespaces from graph.py and main.py

removed trailing whitespaces from main.py

MarvinTorres committed:
Merge branch 'removewhitespaces' into logicchanges

MarvinTorres committed:
Renamed and moved tests

 - test_graph now test_search_results
 - test_graph1, 2, 3 now test_search1_results1, 2, 3
 - fixed test_constrained_path9 in test_search_results1
 - class names changed to reflect new file names
 - moved files to unit folder

MarvinTorres committed:
Fixed padding in graph.py

MarvinTorres committed:
Cleaned up constrained_flexible_paths and added comments

 - compacted default_edge_list and flexible assignments
 - renamed some variables
 - changed loop from for to while

MarvinTorres committed:
Removed failing tests for checking

MarvinTorres committed:
Fixed formatting issues

MarvinTorres committed:
Fixed more formatting issues

MarvinTorres committed:
Minimized local variable count

 - 5 parameters instead of 6
 - 17 local vars instead of 18

MarvinTorres committed:
Fixed all formatting errors

MarvinTorres committed:
Fixed minor formatting error

MarvinTorres committed:
Changed logic to support code reuse

MarvinTorres committed:
Revamped parameter passing.

 - calling constrained flexible paths is now more user-friendly
 - test updated to reflect changes.

MarvinTorres commented:
Okay, I think I have everything. Let me know if I am missing anything crucial.

MarvinTorres committed:
Changed REST API parameters

MarvinTorres committed:
Fixed KytosGraph import in Main

hdiogenes commented on tests/unit/test_graph.py:
```diff
@@ -73,12 +73,13 @@ def test_remove_switch_hops(self):
-        self.assertEqual(shortest_paths, ["any"])
+    #    mock_shortest_simple_paths.assert_called_with(self.kytos_graph.graph,
+    #                                                  source, dest, None)
+    #    self.assertEqual(shortest_paths, ["any"])
```
Why were those two tests commented out? Were they moved elsewhere? If yes, please remove them.

MarvinTorres replied:
We commented them out because they called `_set_default_metadata(...)`, which no longer exists in our implementation of graph.py.

They were not moved. 

I can add the method back, but I will need to research how to do that without impacting our revision history.

hdiogenes replied:
Your revision history will probably be squashed into one big commit before merging. If you want us to maintain it somehow, try to reduce it to just a handful of commits by `git rebase`'ing/squashing it yourself.

MarvinTorres replied:
I have readded the commented-out tests.

hdiogenes commented on main.py:
```diff
@@ -89,7 +89,42 @@ def shortest_path(self):
+                                                          flexible=fle_metrics)
+            return jsonify(paths)
+        except TypeError as err:
+            return jsonify({"error": err})
```
Why `shortest_constrained_path2`? Is there a more descriptive name for this method, instead of 2?
Also, why is it listed as `v4`, when the other is `v3`? Finally, it doesn't look like you need a different method if all you're doing is getting additional (optional) parameters.

MarvinTorres replied:
We added `shortest_constrained_path2` after our product owner suggested that we also let the user select flexible and inflexible metrics. So to implement the alternate design in a timely fashion, we created that method to avoid possibly affecting the implementation of our original design.

But at this point, both methods can be consolidated into one.

hdiogenes commented on main.py:
```diff
@@ -89,7 +89,42 @@ def shortest_path(self):
+        except TypeError as err:
+            return jsonify({"error": err})
+
+    @ listen_to('kytos.topology.updated')
```
There's a space after the `@` symbol.

MarvinTorres replied:
Will fix.

hdiogenes commented on tests/unit/test_search_results.py:
```diff
@@ -0,0 +1,75 @@
+from unittest import TestCase
+
+import networkx as nx
+# Core modules to import
```
Redundant comment.

hdiogenes commented on tests/unit/test_search_results1.py:
```diff
@@ -0,0 +1,135 @@
+"""Module to test the KytosGraph in graph.py."""
+# Core modules to import
```
This comment is not really necessary, and can be safely removed (`from kytos.core... import` says it all ;-)).

hdiogenes commented on tests/unit/test_search_results3.py:
```diff
@@ -0,0 +1,191 @@
+                    elif meta_data and "delay" in meta_data.keys():
+                        delays.append(meta_data["delay"])
+
+        # Assert
```
Please remove all "Arrange / Act / Assert" comments. All unit tests follow the same structure, so they're not needed. It's still recommended keeping the blocks separated by blank lines as you did.

MarvinTorres replied:
Ok, will do. Thank you!

hdiogenes commented on graph.py:
```diff
@@ -46,23 +91,14 @@ def update_links(self, links):
                 self.graph.add_edge(link.endpoint_a.id, link.endpoint_b.id)
                 for key, value in link.metadata.items():
-                    keys.extend(key)
+                    keys.append(key)
```
Are you sure this change is really safe and won't break anything?

MarvinTorres replied:
This is indeed important for me to explain. I will get back to you on this.

MarvinTorres replied:
EDIT: The changed method is not in-scope of our changes, so it will be reverted.

EDIT2: Actually, it seems like it is a bugfix applied early in our development. I will have to find the bug the change was meant to fix.

Original post:
Hello @hdiogenes, I looked at the bodies of the HTTP POST responses after searching for the best paths between the same two points using the current algorithm, which uses extend, and this PR's algorithm, which uses append. They are the same:

Current algorithm results:
![Screenshot from 2020-07-03 14-52-30](https://user-images.githubusercontent.com/26232289/86493210-998d0780-bd3e-11ea-96d2-523fc8cc9055.png)

PR algorithm results:
![Screenshot from 2020-07-03 14-55-14](https://user-images.githubusercontent.com/26232289/86493217-a0b41580-bd3e-11ea-9124-64bf8d681f66.png)

This feels like good news but it does not answer your question, since append and extend work differently and the change might affect something I missed.

I will keep looking into this.

MarvinTorres replied:
I have reverted this change.

EDIT: I will revert the revert, see below.

MarvinTorres replied:
I talked with David. He said that extend was causing each character in the key string to be added to keys instead of the whole string. Changing it to `append` fixed this.

hdiogenes commented on graph.py:
```diff
@@ -46,23 +91,14 @@ def update_links(self, links):
-        for key in keys:
-            for endpoint_a, endpoint_b in self.graph.edges:
-                if key not in self.graph[endpoint_a][endpoint_b]:
-                    self.graph[endpoint_a][endpoint_b][key] = 0
```
Same question here: are you sure this change is really safe and won't break anything?

MarvinTorres replied:
I will check because it is indeed important.

MarvinTorres replied:
I have reverted this removal.

hdiogenes commented on graph.py:
```diff
@@ -10,11 +12,54 @@
+            (int, float), filter_leq("utilization"))
+        self._filter_fun_dict["delay"] = Filter(
+            (int, float), filter_leq("delay"))
+        self._path_fun = nx.all_shortest_paths
```
Suggestion: `path_function`. See comment about abbreviations above.

hdiogenes commented on graph.py:
```diff
@@ -10,11 +12,54 @@
+            (int, float), filter_geq("reliability"))
+        self._filter_fun_dict["utilization"] = Filter(
+            (int, float), filter_leq("utilization"))
+        self._filter_fun_dict["delay"] = Filter(
```
Suggestion: `_filter_dict` or `_filter_functions`.

hdiogenes commented on graph.py:
```diff
@@ -74,9 +110,62 @@ def _remove_switch_hops(circuit):
+
+    def _filter_edges(self, edges, **metrics):
+        for metric, value in metrics.items():
+            fil = self._filter_fun_dict.get(metric, None)
```
Suggestion: `filter_` instead of `fil`.

hdiogenes commented on tests/unit/test_search_results1.py:
```diff
@@ -0,0 +1,135 @@
+"""Module to test the KytosGraph in graph.py."""
```
What's the difference between `test_search_results1`, `test_search_results2` and `test_search_results3`? Can they have more descriptive names?

MarvinTorres replied:
Each file tests a different graph.

I can consolidate them into a single test suite.

EDIT: They actually have more differences. `test_search_results1` tests simple results, such as the case when source == destination or when the destination is unreachable from the source. `test_search_results2` tests if the paths returned have only legal edges. `test_search_results3` tests if the metadata in those edges have passing values.

I will change the names to reflect these differences.

hdiogenes commented on graph.py:
```diff
@@ -10,11 +12,54 @@
+
+    def __init__(self, filter_type, filter_function):
+        self._filter_type = filter_type
+        self._filter_fun = filter_function
```
Please avoid abbreviations. This reads more like "fun filter" than "filter function".

MarvinTorres committed:
Implement Humberto's fixes

ajoaoff commented on tests/unit/test_search_results.py:
```diff
@@ -0,0 +1,75 @@
+        self.graph.clear()
+        self.graph.update_nodes(switches)
+        self.graph.update_links(links)
+        self.graph.set_path_fun(nx.shortest_simple_paths)
```
You are testing using `nx.shortest_simple_paths` as the path function, but in the actual code only `nx.all_shortest_paths` is used. Why are you testing with a different function?

MarvinTorres replied:
`nx.shortest_simple-paths` returns a subset of `nx.all_shortest_paths`, namely the paths that do not have repeating nodes. So it was likely added to save time testing the new KytosGraph implementation.

I removed the call to the setter, so KytosGraph uses `nx.all_shortest_paths` instead, and did not see a regression in the test results.

ajoaoff commented on graph.py:
```diff
@@ -10,11 +12,54 @@
+            (int, float), filter_leq("delay"))
+        self._path_fun = nx.all_shortest_paths
+
+    def set_path_fun(self, path_fun):
+        """Set the shortest path function."""
+        self._path_fun = path_fun
```
This method is only called by tests. Is it really necessary?

MarvinTorres replied:
It is not required and will be removed.

ajoaoff commented on graph.py:
```diff
@@ -74,9 +110,62 @@ def _remove_switch_hops(circuit):
+        # Create "sub-subgraphs" from original subgraph by trimming edges
+        # that fail to meet flexible requirements. Search for a shortest
+        # path in each of these graphs, until at least one is found.
+        while (paths == [] and i in range(0, depth+1)):
+            for combo in combinations(flexible.items(), length-i):
+                additional = dict(combo)
+                paths = self._constrained_shortest_paths(
+                    source, destination, ((u, v) for u, v, d in
+                                          self._filter_edges(default_edge_list,
+                                                             **additional)))
+                if paths != []:
+                    results.append(
+                        {"paths": paths, "metrics": {**base, **additional}})
+            i = i + 1
```
This loop runs in <img src="https://render.githubusercontent.com/render/math?math=O(2^n)" />. Have you stressed this method using a graph with a high number of edges to see if it runs in a reasonable time?

MarvinTorres replied:
Indeed, test_search_results2 uses a graph with 81 edges and runs reasonably quickly, but it could be faster.

I checked the code and found that it is indeed **O(2^n)** but the number of edges does not affect the number of computations. Only the size of the flexible metrics dictionary affects the complexity. I will double-check though.

ajoaoff replied:
Yes, you're right, the combinations are on the metrics, not the edges. My mistake.

ajoaoff commented on tests/unit/test_search_results.py:
```diff
@@ -0,0 +1,75 @@
+class TestSearchResults(TestCase):
+    """Tests for the graph class."""
+
+    def setup(self):
```
```suggestion
    def setUp(self):
```
This way, it is automatically called before each test.

MarvinTorres replied:
Fixed, thank you!

ajoaoff commented on tests/unit/test_search_results1.py:
```diff
@@ -0,0 +1,135 @@
+
+    def test_path1(self):
+        """Tests a simple, definetly possible path"""
+        self.setup()
```
```suggestion
```
This should be automatically called by changing the method's name.

MarvinTorres replied:
Fixed, thank you!

MarvinTorres committed:
Updated comments and changed endpoints for new UI

MarvinTorres committed:
Added missing endpoint changes

MarvinTorres committed:
Updated KytosGraph, TestResults and subclasses
 - KytosGraph
   - set_path_function removed.
 - TestResults
   - test_setup method removed.
   - setup method renamed to setUp to run at the start of every test
 - TestResultsSimple, TestResultsMetaData, TestResultsEdges
   - Removed explicit setup call

MarvinTorres committed:
Gave better descriptions to test subclasses.

MarvinTorres committed:
Undid KytosGraph import change in Main.

MarvinTorres committed:
Linting fixes
 - TestResults
   - removed unused import.
 - TestResultsMetadata
   - fixed two methods with same name.

MarvinTorres committed:
Fixed JSON serialization of error message

MarvinTorres committed:
Re-added methods and reverted update_links

MarvinTorres committed:
Fixed linting issue

MarvinTorres commented:
@hdiogenes @ajoaoff I have updated the original post with a link to the REST API Documentation.

MarvinTorres committed:
changed extend back to append

gleybersonandrade commented on tests/unit/test_results.py:
```diff
@@ -0,0 +1,68 @@
+                                                     maximum_misses,
+                                                     **metrics)
+
+    @ staticmethod
```
@MarvinTorres, can you change these decorators? Just to standardize, thanks!

```suggestion
    @staticmethod
```

MarvinTorres replied:
Okay, will do. Thank you!

MarvinTorres commented:
Hi everyone, I streamlined the history of our changes to prepare for the eventual squash. Make sure that your local repository has the latest version of my `logicchanges` branch.

EDIT: I'm not sure why but updating my remote branch to reflect the streamline broke this PR. Will see if I can fix.

EDIT: Fixed. Turns out that I can not recreate (delete then push) this remote branch without closing this PR. I fixed it by moving `logicchanges` back to the last commit in the original timeline and running:

`git push origin <SHA1 of last commit>:logicchanges --force`

and reopening this PR.

MarvinTorres committed:
Update tests/unit/test_results.py

Standardized decorators

Co-authored-by: Gleyberson Andrade <gleybersonandrade@gmail.com>

MarvinTorres committed:
Updated return values of shortest_constrained_path
   - Returns 400 BAD REQUEST if user provides an illegal attribute value

MarvinTorres commented:
@hdiogenes I updated the REST API link to show the result of rendering #62

hdiogenes commented on tests/unit/test_results.py:
```diff
@@ -0,0 +1,68 @@
+        links = {}
+        return (switches, links)
+
+    @ staticmethod
```
Please do this in all the files where you're using '@ staticmethod' with a space.

```suggestion
    @staticmethod
```

hdiogenes committed:
Remove space from decorator

MarvinTorres commented:
@hdiogenes @gleybersonandrade @ajoaoff I have created an alternate branch **logicchanges_squashed** which contains an identical (changes-wise), linear, and easy-to-squash version of this timeline. Check it out here: https://github.com/MarvinTorres/pathfinder/tree/logicchanges_squashed

EDIT: A separate branch, which serves as the squashed timeline, can be found [here.](https://github.com/MarvinTorres/pathfinder/tree/logicchanges_squashed2)

MarvinTorres committed:
Changed maximum misses variable to minimum hits
   - This is to reflect changes from #62

MarvinTorres committed:
Set up mock graph with metadata

MarvinTorres committed:
Updated testing structure and files
   - test_results and children were moved to integration folder
   - test_graph and test_main now test our added methods
   - test_filter added
   - methods added to test helper

MarvinTorres committed:
Linting fixes

beraldoleal commented:
Hi @MarvinTorres, first of all, thank you for this work. Any improvement is more than welcome.

But, this PR has 100 commits and there is no way to make a good decision on that. Please try to clean up your commit messages and submit individual PRs. The general rule of thumb is that we should not be ashamed of git log history.

beraldoleal reviewed (changes requested):
Hi @MarvinTorres, first of all, thank you for this work. Any improvement is more than welcome.

But, this PR has 100 commits and there is no way to make a good decision on that. Please try to clean up your commit messages and submit individual PRs. The general rule of thumb is that we should not be ashamed of git log history.

MarvinTorres commented:
@beraldoleal Understood. I will create several more PRs that feature bug fixes from our senior project changes.

I have also squashed the 100 changes in a separate branch. [Link](https://github.com/MarvinTorres/pathfinder/tree/logicchanges_squashed2)

beraldoleal commented:
No problem @MarvinTorres.

Just make sure that each commit has a mean full description and the changes make sense for the commit itself. A big squash into a single commit is also not desirable, we need the right balance.

As a suggestion I would recommend applying on PR at the time, to avoid logical dependency and frustration. Sending a PR that depending on another PR could lead to rework and it is not desired.

Also, I haven't looked closely here, but based on your descriptions and diagrams this seems to me that is not a small change/fix. I'm assuming that you have discussed this openly in a blueprint or issue about this before sending such a big change. I will wait for the smaller PR to comment properly on that.

https://docs.kytos.io/developer/how_to_contribute/#submitting-a-pull-request

hdiogenes commented:
Superseded by #68
